### PR TITLE
fix: use destination calendar email

### LINKED ISSUE #3643: [CAL-1684] wrong meeting organizer used with multiple Google accounts
### Issue Summary

I have multiple Google accounts connected (one personal, and two for work):

image

and some event types which are configured to book events onto one of the secondary (i.e. non-default) Google calendars which I use for work:

image

When bookings are made with these event types, they appear correctly on the secondary calendar, but with my cal.com account's main email address as the organizer:

image

### Steps to Reproduce

1. Connect multiple Google accounts (actually this might be reproducible even with one Google account, if that account's email address doesn't match the cal.com account's email address?)
2. Create an event type which books onto a specific Google calendar, and auto-creates a Google Meet link.
3. Get someone to book an event using this event type.
4. Observe that the event organizer is listed as my cal.com account's main email address, rather than my Google account's email address.
5. Observe that despite this, joining the Google Meet from the same work account which owns the calendar it's booked onto results in me correctly being assigned as meeting host.

### Expected results

The calendar event should list my Google work account's email address as the organizer, and not mention the personal email address which my cal.com account is registered with.

## Related issues

I found other issues probably related to this:

- #8444
- #11734 
- #6751

<sub>[CAL-1684](https://linear.app/calcom/issue/CAL-1684/wrong-meeting-organizer-used-with-multiple-google-accounts)</sub>

### LINKED ISSUE #5774: [CAL-499] Choosing the email you want to be displayed as the organizer
**Problem**

* I have a personal [cal.com](http://cal.com) account for (eg [myname@gmail.com](mailto:myname@gmail.com)).
* I also want to use [cal.com](http://cal.com) for my work purposes. To do that I attached my corporate Gmail calendar (eg [myname@mycompany.com](mailto:myname@mycompany.com)) to the existing personal [cal.com](http://cal.com) account. As a result now I get scheduled events in the right calendar ([myname@mycompany.com](mailto:myname@mycompany.com)) but as a participant I am displayed there with my personal email ([myname@gmail.com](mailto:myname@gmail.com))

**Desired outcome**

* Let the user choose on a per event basis which email address for them to appear as the organiser for.
* The list of choices should be pulled from their list of connected calendars.
* We would likely place this alongside the "Add to calendar" UI in the event type settings

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-499](https://linear.app/calcom/issue/CAL-499/choosing-the-email-you-want-to-be-displayed-as-the-organizer)</sub>

### PR DESCRIPTION
**Pull request by Udit-takkar**
Fixes: https://github.com/calcom/cal.com/issues/3643
Fixes: https://github.com/calcom/cal.com/issues/5774

How to test?
1) Connect calendar integration (like google) with email address different than the account
2) Go to Event Type Advanced Settings and turn on the switch and click 'Save'

Screenshot 2024-02-08 at 3 13 00 PM

3) Now Book a meeting on the preview URL

Check if email and calendar event is sent to calendar integration email and not account settings email.

Try different combinations:- outlook integration + team event type.. etc

Preference order:- 
Event type Destination email >  Account Email address

TODO

- [x] Updating all the tests
- [x] Testing all edge cases (Individual event type, team event types)

Could break core features

### CODE REVIEW

**Comment by PeerRich** [comment]
whats needed here?

File: packages/features/bookings/lib/handleNewBooking/test/fresh-booking.test.ts [hunk 1]
```diff
@@ -224,6 +233,7 @@ describe("handleNewBooking", () => {
           organizer,
           emails,
           iCalUID,
+          destinationEmail: organizerDestinationCalendarEmailOnEventType,
```
Comment (Udit-takkar) [inline packages/features/bookings/lib/handleNewBooking/test/fresh-booking.test.ts:236]: Added test here for checking if email is sent to "destination calendar integration" email.

File: packages/prisma/schema.prisma [hunk 2]
```diff
@@ -166,6 +166,8 @@ model DestinationCalendar {
   id           Int         @id @default(autoincrement())
   integration  String
   externalId   String
+  /// @zod.email()
+  primaryEmail String?
```
Comment (Udit-takkar) [inline packages/prisma/schema.prisma:170]: This new field is required because if user has selected any secondary calendar as destination calendar then the externalId is no an actual email. (for eg:- it of format fafasdfsfsaffasdf@group.calendar.google.com )

**Review by joeauyeung — DISMISSED** [review]
Great work so far @Udit-takkar.

It feels like we should give users the option of using either the destination calendar email or their account email. I would connect with product to see where this option should be. I think I saw something in the Figma files.

File: apps/web/pages/booking/[uid].tsx [hunk 3]
```diff
@@ -467,7 +467,9 @@ export default function Success(props: SuccessProps) {
                                 </span>
                                 <Badge variant="blue">{t("Host")}</Badge>
                               </div>
-                              <p className="text-default">{bookingInfo.user.email}</p>
+                              <p className="text-default">
+                                {bookingInfo.destinationCalendar?.primaryEmail ?? bookingInfo.user.email}
+                              </p>
```
[the remark this answers is not in the document]
Comment (Udit-takkar) [reply apps/web/pages/booking/[uid].tsx:472]: Ah yes right. In the future if user removes the destination calendar then /bookings/[uid] would display different email.

**Comment by Udit-takkar** [comment]
@joeauyeung 

> It feels like we should give users the option of using either the destination calendar email or their account email. I would connect with product to see where this option should be. I think I saw something in the Figma files.

Yeah there was a design earlier in figam ("settings/my-account/profile" page). There was an option to add another email. maybe we add the option there to select email to be used for sending emails and calendar events .. etc.

**Comment by joeauyeung** [comment]
> @joeauyeung
> 
> > It feels like we should give users the option of using either the destination calendar email or their account email. I would connect with product to see where this option should be. I think I saw something in the Figma files.
> 
> Yeah there was a design earlier in figam ("settings/my-account/profile" page). There was an option to add another email. maybe we add the option there to select email to be used for sending emails and calendar events .. etc.

I would feel comfortable shipping this once we give users that option.

**Comment by Udit-takkar** [comment]
Found the designs https://www.figma.com/file/UajFu4M1APhn2AywAEJkJr/New-Features?node-id=1021%3A104990&mode=dev

File: apps/web/pages/booking/[uid].tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (joeauyeung) [reply apps/web/pages/booking/[uid].tsx:472]: Thoughts on this @ciaranha @itsberkaya? I'm leaning towards keeping the email static with the booking information but wanted to get your thoughts.

File: apps/web/pages/booking/[uid].tsx [hunk 4]
```diff
@@ -1115,6 +1117,7 @@ export async function getServerSideProps(context: GetServerSidePropsContext) {
       cancellationReason: true,
       responses: true,
       rejectionReason: true,
+      destinationCalendar: true,
```
Comment (zomars) [inline apps/web/pages/booking/[uid].tsx:1120]:
```suggestion
      destinationCalendar: { select: { primaryEmail: true } },
```

Comment (zomars) [reply apps/web/pages/booking/[uid].tsx:1120]: Let's keep the payload at minimum.

File: apps/web/pages/booking/[uid].tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (zomars) [reply apps/web/pages/booking/[uid].tsx:472]: Makes sense to "capture" the email on the booking itself. That way it can retain historical value.

**Comment by ciaranha** [comment]
@Udit-takkar Are you blocked in this?

File: packages/prisma/schema.prisma [hunk 5]
```diff
@@ -451,6 +454,9 @@ model Booking {
   uid                   String               @unique
   user                  User?                @relation(fields: [userId], references: [id], onDelete: Cascade)
   userId                Int?
+  // User's email at the time of booking
+  /// @zod.email()
+  userPrimaryEmail      String?
```
Comment (Udit-takkar) [inline packages/prisma/schema.prisma:459]: We need to store only user's email at the time of booking so that even when user changes email or connect another calendar we display the email at the time of booking on /bookings/[uid] and use that

File: apps/web/components/eventtype/EventAdvancedTab.tsx [hunk 6]
```diff
@@ -172,6 +177,18 @@ export const EventAdvancedTab = ({ eventType, team }: Pick<EventTypeSetupProps,
...
+            <Switch
+              label={t("display_add_to_calendar_organizer")}
+              checked={useEventTypeDestinationCalendarEmail}
+              onCheckedChange={(val) => {
+                setUseEventTypeDestinationCalendarEmail(val);
+                formMethods.setValue("useEventTypeDestinationCalendarEmail", val);
+              }}
+            />
+          </div>
+        )}
```
Comment (Udit-takkar) [inline apps/web/components/eventtype/EventAdvancedTab.tsx:191]: Only for individual event types

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 7]
```diff
@@ -1559,7 +1568,10 @@ async function handler(
     organizer: {
       id: organizerUser.id,
       name: organizerUser.name || "Nameless",
-      email: organizerUser.email || "Email-less",
+      email:
+        eventType.useEventTypeDestinationCalendarEmail && destinationCalendar?.[0]?.primaryEmail
+          ? destinationCalendar[0].primaryEmail
+          : organizerUser.email || "Email-less",
```
Comment (Udit-takkar) [inline packages/features/bookings/lib/handleNewBooking.ts:1574]: use the destination email as organizer email only when useEventTypeDestinationCalendarEmail is true

File: packages/prisma/schema.prisma (hunk 5 above)
Comment (Udit-takkar) [reply packages/prisma/schema.prisma:459]: For attendees we save that info in Attendee table and that is not updated even when any attendee changes email as  that new info is stored in User Table

**Review by CarinaWolli — CHANGES_REQUESTED** [review]
We need to make sure to use the correct email in all booking-related emails, I found some places where it's wrong:
- Cancellation email 
- Request reschedule email 
- I didn't test payments but do we also need to change the email in https://github.com/calcom/cal.com/blob/fix/destination-calendar/packages/lib/payment/getBooking.ts#L99 which is called in `handlePaymentSuccess`?

**Comment by Udit-takkar** [comment]
> We need to make sure to use the correct email in all booking-related emails, I found some places where it's wrong:
> 
> * Cancellation email
> * Request reschedule email
> * I didn't test payments but do we also need to change the email in https://github.com/calcom/cal.com/blob/fix/destination-calendar/packages/lib/payment/getBooking.ts#L99 which is called in `handlePaymentSuccess`?

@CarinaWolli Fixed Now. Thanks for the review.

**Review by CarinaWolli — APPROVED** [review]
Looks good now 🙌🏻

**Comment by rthaut** [comment]
Note sure if this is the right place to post this feedback, but neither [CAL-1684] (#3643) nor [CAL-499] (#5774) appear to be working for me on cal.com after enabling the new setting in my Event Types.

- The calendar invite still has my cal.com account's email address in the `ORGANIZER` property, not the email address of the target calendar's owner
    - Verified by viewing the .ics file attached to the email
- The body of the email still has my cal.com account's email address in the "Who" section as the organizer, not the email address of the target calendar's owner
- The email was sent to my cal.com account's email address, not the email address of the target calendar's owner

I have verified the new "Display 'Add to calendar' email as the organizer" toggle is checked on my event types, and I have tried toggling it on and off multiple times.

All calendars are Google Calendars using Google Workspace accounts.

**Comment by CarinaWolli** [comment]
@rthaut It works for me as expected I noticed that the email for your calendar matches the one on your Cal account. Have you made any recent changes, or what's the issue you're trying to resolve?

**Comment by rthaut** [comment]
@CarinaWolli is there a support mechanism other than GitHub tickets for working through this? I don't want to clutter up this PR if it is simply a misconfiguration issue on my part...

> @rthaut I noticed that the email for your calendar matches the one on your Cal account. 

I don't understand how this could be. I have chosen my Event Types to go to calendars that are owned by other email addresses and are simply shared (via Google's Calendar sharing functionality) to the email address I use for my cal.com account.

> Have you made any recent changes?

After your response I did try re-adding my target calendars directly (i.e. I gave cal.com access to each calendar by logging in with the appropriate email address), and updating the "Add to calendar" setting of the Event Types. However, even with that change all invites and emails are still just using my cal.com email address.

> What's the issue you're trying to resolve?

I want certain Event Type links to create appointments on certain calendars. Additionally, I would like to avoid exposing the email address I use for cal.com to anyone scheduling events with me, and instead expose the email addresses associated those to the calendars.

My use case is that I sub-contract for several companies, most of whom provide me with a Google workspace account for their domain. Through those companies I have multiple clients who I work with directly on behalf of the companies for whom I sub-contract. I want all of my clients to see my availability in one place, but I want them to schedule to the calendar/domain to which they are a client. I especially don't want those clients to have any way of finding my private email address/domain.

**Comment by CarinaWolli** [comment]
@rthaut Sounds specific, would be best to reach out to our support [go.cal.com/support](https://go.cal.com/support). And make sure to add all this information 🙏

**Comment by Udit-takkar** [comment]
@rthaut just make sure to delete any old calendar integration and connect them again and then turn on the switch in the event type advanced settings. You would receive all the emails and events on the selected calendar and on the booking page also it would not mention your account's email
